storage/replica_command.go

```diff
@@ -294,27 +312,67 @@ func (r *Replica) EndTransaction(batch engine.Engine, ms *engine.MVCCStats, args
 		// retry error if the commit timestamp isn't equal to the txn
 		// timestamp.
 		if args.Txn.Isolation == proto.SERIALIZABLE && !reply.Txn.Timestamp.Equal(args.Txn.OrigTimestamp) {
-			return reply, proto.NewTransactionRetryError(reply.Txn)
+			return reply, nil, proto.NewTransactionRetryError(reply.Txn)
 		}
 		reply.Txn.Status = proto.COMMITTED
 	} else {
 		reply.Txn.Status = proto.ABORTED
 	}
 
-	// Persist the transaction record with updated status (& possibly timestamp).
-	if err := engine.MVCCPutProto(batch, ms, key, proto.ZeroTimestamp, nil, reply.Txn); err != nil {
-		return reply, err
+	// Resolve any explicit intents. All that are local to this range get
+	// resolved synchronously in the same batch. The remainder is collected
+	// and handed off to asynchronous processing.
+	var externalIntents []proto.Intent
+	desc := *(r.Desc())
+	if wideDesc := args.GetInternalCommitTrigger().GetMergeTrigger().GetUpdatedDesc(); wideDesc.RangeID != 0 {
+		// If this is a merge, then use the post-merge descriptor to determine
+		// which intents are local (note that for a split, we want to use the
+		// pre-split one instead because it's larger).
+		desc = wideDesc
 	}
-
-	// Resolve any explicit intents.
 	for _, intent := range args.Intents {
-		if log.V(1) {
-			log.Infof("resolving intent at %s on end transaction [%s]", intent.Key, reply.Txn.Status)
-		}
-		if err := engine.MVCCResolveWriteIntent(batch, ms, intent.Key, reply.Txn.Timestamp, reply.Txn); err != nil {
-			return reply, err
+		if err := func() error {
+			if len(intent.EndKey) == 0 {
+				// For single-key intents, do a KeyAddress-aware check of
+				// whether it's contained in our Range.
+				if !containsKey(desc, intent.Key) {
+					externalIntents = append(externalIntents, intent)
+					return nil
+				}
+				return engine.MVCCResolveWriteIntent(batch, ms,
+					intent.Key, reply.Txn.Timestamp, reply.Txn)
+			}
+			// For intent ranges, cut into parts inside and outside our key
+			// range. Resolve locally inside, delegate the rest. In particular,
+			// an intent range for range-local data is correctly considered local.
+			// TODO(tschottdorf): This will panic for ranges which contain both
```

Comment:
Why not add a test to storage/store.go in verifyKeys to make sure there's no cross over between non-local and local keyspace. That should be disallowed for sure.